src/Microsoft.ComponentDetection.Detectors/pip/IPyPiClient.cs

```diff
@@ -45,52 +53,68 @@ public class PyPiClient : IPyPiClient
+        /// The response from PyPi is not automatically added to the cache, to allow caller to make that decision.
         /// </summary>
         /// <param name="uri">The REST Uri to call.</param>
-        /// <returns>The cached response or a new result from Pypi.</returns>
-        private async Task<HttpResponseMessage> GetPypiResponse(string uri)
+        /// <returns>The cached response or a new result from PyPi.</returns>
+        private async Task<HttpResponseMessage> GetPyPiResponse(string uri)
         {
-            if (cachedResponses.TryGetValue(uri, out var value))
+            if (cachedResponses.TryGetValue(uri, out HttpResponseMessage result))
             {
-                return await value;
+                Logger.LogVerbose("Retrieved cached Python data from " + uri);
+                return result;
             }
 
             Logger.LogInfo("Getting Python data from " + uri);
             return await HttpClient.GetAsync(uri);
         }
 
         /// <summary>
-        /// Used to update the consistency cache, decision has to be made by the caller to allow for retries!.
+        /// Used to update the in-memory LRU cache, decision has to be made by the caller to allow for retries.
         /// </summary>
         /// <param name="uri">The REST Uri to call.</param>
         /// <param name="message">The proposed response by the caller to store for this Uri.</param>
         /// <returns>The `first-wins` response accepted into the cache.
-        /// This might be different from the input if another caller wins the race!.</returns>
-        private async Task<HttpResponseMessage> CachePypiResponse(string uri, HttpResponseMessage message)
+        /// This might be different from the input if another caller wins the race.</returns>
+        private async Task<HttpResponseMessage> CachePyPiResponseMemory(string uri, HttpResponseMessage message)
         {
-            if (!cachedResponses.TryAdd(uri, Task.FromResult(message)))
+            // On the initial caching attempt, see if the user specified an override
+            // for PyPiMaxCacheEntries and recreate the cache if needed
+            if (!checkedMaxEntriesVariable)
             {
-                return await cachedResponses[uri];
+                var maxEntriesVariable = EnvironmentVariableService.GetEnvironmentVariable("PyPiMaxCacheEntries");
+                if (!string.IsNullOrEmpty(maxEntriesVariable) && long.TryParse(maxEntriesVariable, out var maxEntries))
+                {
+                    Logger.LogInfo($"Setting IPyPiClient max cache entries to {maxEntries}");
+                    cachedResponses = new MemoryCache(new MemoryCacheOptions { SizeLimit = maxEntries });
+                }
+
+                checkedMaxEntriesVariable = true;
             }
 
-            return message;
+            return await cachedResponses.GetOrCreateAsync(uri, cacheEntry =>
```

Comment:
The order of things is a little weird here.

In `GetPyPiResponse` we try to fetch from cache and fallback to making an HTTP call. Here we attempt to fetch from cache again, and only then set.

It might make sense to combine both `GetPyPiResponse` and `CachePyPiResponseMemory` into one?

Reply:
I updated the logic so caching occurs in `GetAndCachePyPiResponse` rather than relying on the caller, since both occurrences were caching it anyway.